pkg/executor/sortexec/sort.go

```diff
@@ -52,42 +56,76 @@ type SortExec struct {
 	memTracker  *memory.Tracker
 	diskTracker *disk.Tracker
 
-	// sortPartitions is the chunks to store row values for partitions. Every partition is a sorted list.
-	sortPartitions []*sortPartition
+	IsUnparallel bool
+
+	finishCh              chan struct{}
+	isFinishChannelClosed *atomic.Bool
+
+	Unparallel struct {
+		Idx int
+
+		// sortPartitions is the chunks to store row values for partitions. Every partition is a sorted list.
+		sortPartitions []*sortPartition
+
+		// multiWayMerge uses multi-way merge for spill disk.
+		// The multi-way merge algorithm can refer to https://en.wikipedia.org/wiki/K-way_merge_algorithm
+		multiWayMerge *multiWayMerge
+
+		// spillAction save the Action for spill disk.
+		spillAction *sortPartitionSpillDiskAction
+	}
 
-	// multiWayMerge uses multi-way merge for spill disk.
-	// The multi-way merge algorithm can refer to https://en.wikipedia.org/wiki/K-way_merge_algorithm
-	multiWayMerge *multiWayMerge
-	// spillAction save the Action for spill disk.
-	spillAction *sortPartitionSpillDiskAction
+	Parallel struct {
+		result sortedRows
+		rowNum int64
+		idx    int64
+
+		chunkChannel    chan *chunk.Chunk
+		isChannelClosed *atomic.Bool
+		workers         []*parallelSortWorker
+
+		// All workers' sorted rows will be put into this list to be merged
+		globalSortedRowsQueue *sortedRowsList
+
+		errRWLock *sync.RWMutex
+		err       error
+	}
 
 	enableTmpStorageOnOOM bool
 }
 
 // Close implements the Executor Close interface.
 func (e *SortExec) Close() error {
-	for _, partition := range e.sortPartitions {
-		err := partition.close()
-		if err != nil {
-			return err
+	if e.Unparallel.spillAction != nil {
+		e.Unparallel.spillAction.SetFinished()
+	}
+
+	if e.IsUnparallel {
+		for _, partition := range e.Unparallel.sortPartitions {
+			partition.close()
+		}
+	} else {
+		if e.Parallel.err != nil {
+			return e.Parallel.err
 		}
+		e.tryToCloseChunkChannel()
 	}
-	e.sortPartitions = e.sortPartitions[:0]
 
-	e.memTracker = nil
-	e.diskTracker = nil
-	e.multiWayMerge = nil
-	if e.spillAction != nil {
-		e.spillAction.SetFinished()
+	if e.memTracker != nil {
+		e.memTracker.ReplaceBytesUsed(0)
```

Comment:
Should cleanup chunkChannel and global_queue before `memTracker.ReplaceBytesUsed(0)`
https://github.com/pingcap/tidb/blob/245951f07d5767047b1b67cf2f9d1ef17003a84d/pkg/executor/projection.go#L323-L324